services/grpc/src/main/java/mil/nga/giat/geowave/service/grpc/GeoWaveGrpcServer.java

```diff
@@ -0,0 +1,333 @@
+					}
+				});
+	}
+
+	/** Stop serving requests and shutdown resources. */
+	public void stop() {
+		if (server != null) {
+			server.shutdown();
+		}
+	}
+
+	/**
+	 * Await termination on the main thread since the grpc library uses daemon
+	 * threads.
+	 */
+	private void blockUntilShutdown()
+			throws InterruptedException {
+		if (server != null) {
+			server.awaitTermination();
+		}
+	}
+
+	private static class GeoWaveGrpcVectorQueryService extends
+			VectorQueryGrpc.VectorQueryImplBase
+	{
+		@Override
+		public void cqlQuery(
+				CQLQueryParameters request,
+				StreamObserver<Feature> responseObserver ) {
+
+			final String cql = request.getCql();
+			final String storeName = request.getBaseParams().getStoreName();
+			final StoreLoader storeLoader = new StoreLoader(
+					storeName);
+
+			final ByteArrayId adapterId = new ByteArrayId(
+					request.getBaseParams().getAdapterId().toByteArray());
+			final ByteArrayId indexId = new ByteArrayId(
+					request.getBaseParams().getIndexId().toByteArray());
+
+			// first check to make sure the data store exists
+			if (!storeLoader.loadFromConfig(ConfigOptions.getDefaultPropertyFile())) {
+				throw new ParameterException(
+						"Cannot find store name: " + storeLoader.getStoreName());
+			}
+
+			DataStore dataStore = null;
+			AdapterStore adapterStore = null;
+			GeotoolsFeatureDataAdapter adapter = null;
+
+			// get a handle to the data store and also the adapter store
+			dataStore = storeLoader.createDataStore();
+			adapterStore = storeLoader.createAdapterStore();
+
+			if (adapterId != null) {
+				adapter = (GeotoolsFeatureDataAdapter) adapterStore.getAdapter(adapterId);
+			}
+
+			try (final CloseableIterator<SimpleFeature> iterator = dataStore.query(
+					new QueryOptions(
+							adapterId,
+							indexId),
+					CQLQuery.createOptimalQuery(
+							cql,
+							adapter,
+							null))) {
+
+				while (iterator.hasNext()) {
+					SimpleFeature simpleFeature = iterator.next();
+					SimpleFeatureType type = simpleFeature.getType();
+					Feature.Builder b = Feature.newBuilder();
+					for (int i = 0; i < type.getAttributeDescriptors().size(); i++) {
+						b.putAttributes(
+								type.getAttributeDescriptors().get(
+										i).getLocalName(),
+								simpleFeature.getAttribute(i) == null ? "" : simpleFeature.getAttribute(
+										i).toString());
+					}
+					Feature f = b.build();
+					responseObserver.onNext(f);
+				}
+				responseObserver.onCompleted();
+			}
+			catch (final Exception e) {
+				LOGGER.error(e.getMessage());
+			}
+		}
+
+		public void spatialQuery(
+				SpatialQueryParameters request,
+				StreamObserver<Feature> responseObserver ) {
+
+			final String storeName = request.getBaseParams().getStoreName();
+			final StoreLoader storeLoader = new StoreLoader(
+					storeName);
+
+			final ByteArrayId adapterId = new ByteArrayId(
+					request.getBaseParams().getAdapterId().toByteArray());
+			final ByteArrayId indexId = new ByteArrayId(
```

Comment:
indexID and adapter ID are typically optional params on a query

Reply:
Leaving these parameters as discussed. They are indeed optional as well when invoking the gRPC call.